gcloud/datastore/batch.py

```diff
@@ -0,0 +1,170 @@
+        :param dataset_id: The ID of the dataset.
+
+        :type connection: :class:`gcloud.datastore.connection.Connection`
+        :param connection: The connection used to connect to datastore.
+
+        :raises: :class:`ValueError` if either a connection or dataset ID
+                 are not set.
+        """
+        self._connection = connection or _implicit_environ.CONNECTION
+        self._dataset_id = dataset_id or _implicit_environ.DATASET_ID
+
+        if self._connection is None or self._dataset_id is None:
+            raise ValueError('A batch must have a connection and '
+                             'a dataset ID set.')
+
+        self._id = None
+        self._mutation = datastore_pb.Mutation()
+
+    @property
+    def dataset_id(self):
+        """Getter for dataset ID in which the batch will run.
+
+        :rtype: :class:`str`
+        :returns: The dataset ID in which the batch will run.
+        """
+        return self._dataset_id
+
+    @property
+    def connection(self):
+        """Getter for connection over which the batch will run.
+
+        :rtype: :class:`gcloud.datastore.connection.Connection`
+        :returns: The connection over which the batch will run.
+        """
+        return self._connection
+
+    @property
+    def mutation(self):
+        """Getter for the current mutation.
+
+        Every batch is committed with a single Mutation
+        representing the 'work' to be done as part of the batch.
+        Inside a batch, calling ``batch.put()`` with an entity, or
+        ``batch.delete`` with a key, builds up the mutation.
+        This getter returns the Mutation protobuf that
+        has been built-up so far.
+
+        :rtype: :class:`gcloud.datastore.datastore_v1_pb2.Mutation`
+        :returns: The Mutation protobuf to be sent in the commit request.
+        """
+        return self._mutation
+
+    def put(self, entity):
+        """Remember an entity's state to be saved during ``commit``.
+
+        .. note::
+           Any existing properties for the entity will be replaced by those
+           currently set on this instance.  Already-stored properties which do
+           not correspond to keys set on this instance will be removed from
+           the datastore.
+
+        .. note::
+           Property values which are "text" ('unicode' in Python2, 'str' in
+           Python3) map to 'string_value' in the datastore;  values which are
+           "bytes" ('str' in Python2, 'bytes' in Python3) map to 'blob_value'.
+
+        :type entity: :class:`gcloud.datastore.entity.Entity`
+        :param entity: the entity to be saved.
+
+        :raises: ValueError if entity has no key assigned.
+        """
+        if entity.key is None:
+            raise ValueError("Entity must have a key")
+
+        key_pb = entity.key.to_protobuf()
+        properties = dict(entity)
+        exclude = tuple(entity.exclude_from_indexes)
+
+        self.connection.save_entity(
```

Comment:
We pass the 'mutation' object in:  in that caes, as when there is a transaction active, the connection does not do the RPC:  it merely does the "copy to mutation" stuff.

One might arguably factor out the "copy to mutation" (into `helpers'?) from `Connection.save_entity` and `Connection.delete_entities`.